### PR TITLE
Added support for protocol 1.12

### LINKED ISSUE #3750: Unsupported Protocol 335
Client version:   1.12
Server OS:        Linux

When I will connect my own server after the latest update the client showes me that he cant connect to my server because the protocol 335 is unsupported and I should use one of these versions: 1.8,1.9.,1.10, 1.11

### PR DESCRIPTION
**Pull request by Seadragon91**
Closes #3750 

This adds support for protocol 1.12. Contains the changes for the packet types and the protocol version number.
Note: This only adds the protocol support, NOT any of the new features that exists in 1.12.

Problems detected:
* ~~Authentication not working, I always switch that off locally...~~
* ~~Can see the ground in water~~
* ~~Leaving a vehicle will close the connection to the server, with an index-out-of-bounds exception~~

### CODE REVIEW

**Comment by Seadragon91** [comment]
Authentication does work, some problem with my vm machine.

**Comment by Seadragon91** [comment]
Fixed disconnect problem when leaving a vehicle.

File: src/Protocol/ProtocolRecognizer.h [hunk 1]
```diff
@@ -18,9 +18,9 @@
 
 
 // Adjust these if a new protocol is added or an old one is removed:
-#define MCS_CLIENT_VERSIONS "1.8.x, 1.9.x, 1.10.x, 1.11.x"
-#define MCS_PROTOCOL_VERSIONS "47, 107, 108, 109, 110, 210, 315, 316"
-#define MCS_LATEST_PROTOCOL_VERSION 316
+#define MCS_CLIENT_VERSIONS "1.8.x, 1.9.x, 1.10.x, 1.11.x 1.12"
```
Comment (madmaxoft) [inline src/Protocol/ProtocolRecognizer.h]: Comma missing

File: src/Protocol/Protocol_1_12.cpp [hunk 2]
```diff
@@ -0,0 +1,1509 @@
...
+			// Game
+			switch (a_PacketType)
+			{
+				case 0x00: HandleConfirmTeleport(a_ByteBuffer); return true;
+				case 0x01: break;  // Prepare Crafting Grid, not yet implemented
+				case 0x02: HandlePacketTabComplete(a_ByteBuffer); return true;
+				case 0x03: HandlePacketChatMessage(a_ByteBuffer); return true;
+				case 0x04: HandlePacketClientStatus(a_ByteBuffer); return true;
+				case 0x05: HandlePacketClientSettings(a_ByteBuffer); return true;
+				case 0x06: break;  // Confirm transaction - not used in MCS
```
Comment (madmaxoft) [inline src/Protocol/Protocol_1_12.cpp]: `MCS` -> `Cuberite` (possibly elsewhere, too, didn't check all)

File: src/Protocol/Protocol_1_12.h [hunk 3]
```diff
@@ -0,0 +1,66 @@
+
+// Protocol_1_11.h
```
Comment (madmaxoft) [inline src/Protocol/Protocol_1_12.h]: Bad number

File: src/Protocol/Protocol_1_12.h [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+/*
+Declares the 1.11 protocol classes:
```
Comment (madmaxoft) [inline src/Protocol/Protocol_1_12.h]: Bad number

File: src/Protocol/Protocol_1_12.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	- cProtocol_1_12
+		- release 1.11 protocol (#315)
+(others may be added later in the future for the 1.11 release series)
```
Comment (madmaxoft) [inline src/Protocol/Protocol_1_12.h]: Bad number (2x)

**Comment by Seadragon91** [comment]
I see the ground in 1.11 too.

**Comment by Bond-009** [comment]
For me compiling fails after the merge of this PR on macOS.

`Apple LLVM version 8.1.0 (clang-802.0.42)`
```
cuberite/src/Protocol/Protocol_1_12.h:35:15: error: 'SendDetachEntity' overrides a
      member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]
        virtual void SendDetachEntity(const cEntity & a_Entity, const cEntity & a_PreviousVehicle);
                     ^
cuberite/src/Protocol/Protocol_1_9.h:80:15: note: overridden virtual function is
      here
        virtual void SendDetachEntity               (const cEntity & a_Entity, const cEntity & a_PreviousVehicle) override;
```

**Comment by Seadragon91** [comment]
Will fix it.

**Comment by madmaxoft** [comment]
I'm doing it already

**Pull request merged**